compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -289,16 +293,16 @@ pub trait TypeErrCtxtExt<'tcx> {
         next_code: Option<&ObligationCauseCode<'tcx>>,
     );
 
-    fn note_obligation_cause_code<T>(
+    fn note_obligation_cause_code(
         &self,
         err: &mut Diagnostic,
-        predicate: &T,
+        predicate: Option<ty::Predicate<'tcx>>,
+        predicate_fallback: String,
```

Comment:
#105342 should fix this strange wart by making the `predicate` argument always impl `ToPredicate`.

Reply:
I wanted to, but I would have to add a variant for the case of a Ty Predicate :-/

Reply:
The PR I linked avoids that by not noting the cause code for `Obligation<Ty<'tcx>>` -- the only case that happens is during opaque type overflow, which only happens during codegen, and is almost certainly not associated with a useful obligation cause.

Reply:
Ah, great. If that works great (but I was seeing real cases of type overflows with a need to display useful obligation causes. I'll rebase on top of yours after it lands.

Reply:
`<T as ToPredicate>::to_predicate` should give you back a ty::Predicate, not a `&T`?